ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs

```diff
@@ -0,0 +1,364 @@
+            | sibling ->
+                find sibling
+
+        find node
+
+    let getTypeUsage (pattern: IFSharpPattern) =
+        findSibling<ITypeUsage> pattern
+
+    let getReturnTypeInfo (pattern: IFSharpPattern) =
+        findSibling<IReturnTypeInfo> pattern
+
+    let removeTypeAnnotations (pattern: IFSharpPattern) =
+        match pattern with
+        | :? ITypedPat as typedPat ->
+            ModificationUtil.ReplaceChild(typedPat, typedPat.Pattern.IgnoreInnerParens())
+        | _ ->
+            match getTypeUsage pattern with
+            | ValueSome typeUsage ->
+                ModificationUtil.DeleteChild(typeUsage)
+            | _ ->
+                match getReturnTypeInfo pattern with
+                | ValueSome typeInfo ->
+                    ModificationUtil.DeleteChild(typeInfo)
+                | _ ->
+                    ()
+
+            let colon = pattern.GetNextMeaningfulSibling()
+            if isNotNull colon && colon.GetTokenType() == FSharpTokenType.COLON then
+                ModificationUtil.DeleteChild(colon)
+
+            pattern
+
+    let removeInnerParens (pattern: IFSharpPattern) =
+        let updatedPattern = pattern.IgnoreInnerParens()
+        if pattern == updatedPattern then
+            pattern
+        else
+            ModificationUtil.ReplaceChild(pattern, updatedPattern)
+
+module FcsMfvUtil =
+    let getFunctionReturnType parameters (fcsType: FSharpType) =
+        let rec skipFunctionParameters remaining (fcsType: FSharpType) =
+            if remaining = 0 then fcsType
+            else
+                skipFunctionParameters (remaining - 1) fcsType.GenericArguments[1]
+
+        let returnType = skipFunctionParameters parameters fcsType
+        returnType
+
+    let getFunctionParameterTypes parameters (fcsType: FSharpType) =
+        let result = Array.zeroCreate parameters
+        let mutable fullType = fcsType
+
+        for i = 0 to parameters - 1 do
+            result[i] <- fullType.GenericArguments[0]
+            fullType <- fullType.GenericArguments[1]
+
+        result
+
+    // methods take "this" arg as first parameter so skip it
+    let getMethodParameterTypes parameters (fcsType: FSharpType) =
+        let result = Array.zeroCreate parameters
+        let mutable fullType = fcsType.GenericArguments[1]
+
+        for i = 0 to parameters - 1 do
+            result[i] <- fullType.GenericArguments[0]
+            fullType <- fullType.GenericArguments[1]
+
+        result
+
+    let getFunctionParameterAt parameterIndex (fcsType: FSharpType) =
+        let rec getParameter index (fcsType: FSharpType) =
+            if index = 0 then fcsType.GenericArguments[0]
+            else
+                getParameter (index - 1) fcsType.GenericArguments[1]
+
+        getParameter parameterIndex fcsType
+
+module AnnotationUtil =
+
+    let private isFullyAnnotatedNamedTypeUsage (namedTypeUsage: INamedTypeUsage) =
+        isNull namedTypeUsage.ReferenceName ||
```

Comment:
Not critical: this looks a bit counter-intuitive that not having a name makes it considered as fully annotated. Maybe a name like `needsAnnotation` could make it clearer.

Reply:
I've changed this a little, please take a look